jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java

```diff
@@ -190,633 +89,348 @@ public Interceptor getInterceptor()
+                if (_contentProducer.getRawContentArrived() < minimumData)
+                {
+                    BadMessageException bad = new BadMessageException(HttpStatus.REQUEST_TIMEOUT_408,
+                        String.format("Request content data rate < %d B/s", minRequestDataRate));
+                    if (_channelState.isResponseCommitted())
+                        _channelState.getHttpChannel().abort(bad);
+                    throw bad;
+                }
             }
-            return false;
         }
-        catch (IOException e)
-        {
-            LOG.ignore(e);
-            return true;
-        }
-    }
 
-    @Override
-    public void setReadListener(ReadListener readListener)
-    {
-        boolean woken = false;
-        try
+        while (true)
         {
-            synchronized (_inputQ)
+            _semaphore.drainPermits();
+            int read = _contentProducer.read(this::produceRawContent, b, off, len);
+            if (LOG.isDebugEnabled())
+                LOG.debug("read produced {} byte(s)", read);
+            if (read > 0)
+                return read;
+
+            if (LOG.isDebugEnabled())
+                LOG.debug("read error = " + _error);
+            if (_error != null)
+                throw new IOException(_error);
+
+            if (LOG.isDebugEnabled())
+                LOG.debug("read EOF = {}", _eof);
+            if (_eof.isEarly())
+                throw new EofException("Early EOF");
+
+            if (LOG.isDebugEnabled())
+                LOG.debug("read async = {}", isAsync());
+            if (!isAsync())
             {
-                if (_listener != null)
-                    throw new IllegalStateException("ReadListener already set");
-
-                _listener = Objects.requireNonNull(readListener);
-
-                if (isError())
+                if (_eof.isEof())
+                {
+                    _eof = Eof.CONSUMED_EOF;
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("read on EOF, switching to CONSUMED_EOF and returning");
+                    return -1;
+                }
+                if (LOG.isDebugEnabled())
+                    LOG.debug("read blocked");
+                blockForContent();
+                if (LOG.isDebugEnabled())
+                    LOG.debug("read unblocked");
+            }
+            else
+            {
+                if (_eof.isEof())
                 {
-                    woken = _channelState.onReadReady();
+                    boolean wasInAsyncWait = _channelState.onReadEof();
+                    if (wasInAsyncWait)
+                        scheduleReadListenerNotification();
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("async read on EOF (was in async wait? {}), switching to CONSUMED_EOF and returning", wasInAsyncWait);
+                    _eof = Eof.CONSUMED_EOF;
```

Comment:
This assignment needs to be done before the scheduleReadListenerNotification to avoid a race.

Reply:
I *think* it's safe the way it is, at least for now. But I do agree that the assignment should be done before the `scheduleReadListenerNotification()` call.